File: Scala/0017.scala
object Solution {
    def letterCombinations(digits: String): List[String] = {
        // Mapping of digits to letters as per the phone keypad
        val digitToLetters = Map(
            '2' -> "abc",
            '3' -> "def",
            '4' -> "ghi",
            '5' -> "jkl",
            '6' -> "mno",
            '7' -> "pqrs",
            '8' -> "tuv",
            '9' -> "wxyz"
        )

        // Initialize result list
        val results = scala.collection.mutable.ListBuffer[String]()

        // Helper function to perform backtracking
        def backtrack(index: Int, path: StringBuilder): Unit = {
            if (index == digits.length) {
                // If we have processed all digits, add the path to results
                results += path.toString
            } else {
                val digit = digits(index)
                val letters = digitToLetters.getOrElse(digit, "")
                for (letter <- letters) {
                    path.append(letter)
                    backtrack(index + 1, path)
                    path.deleteCharAt(path.length - 1) // Backtrack
                }
            }
        }

        if (digits.nonEmpty) {
            backtrack(0, new StringBuilder)
        }

        results.toList
    }
}

File: C/0017.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Define the mapping from digits to letters
const char *digitToChar[8] = {
    "abc",  // for digit '2'
    "def",  // for digit '3'
    "ghi",  // for digit '4'
    "jkl",  // for digit '5'
    "mno",  // for digit '6'
    "pqrs", // for digit '7'
    "tuv",  // for digit '8'
    "wxyz"  // for digit '9'
};

// Helper function for backtracking
void backtrack(char *combination, int index, char *digits, char **result, int *returnSize) {
    if (index == strlen(digits)) {
        result[*returnSize] = strdup(combination);
        (*returnSize)++;
        return;
    }

    int digit = digits[index] - '2';
    const char *letters = digitToChar[digit];
    
    for (int i = 0; letters[i] != '\0'; i++) {
        combination[index] = letters[i];
        backtrack(combination, index + 1, digits, result, returnSize);
    }
}

// Main function to return letter combinations
char** letterCombinations(char* digits, int* returnSize) {
    *returnSize = 0;
    if (strlen(digits) == 0) {
        return NULL;
    }

    int n = strlen(digits);
    int totalCombinations = 1;

    // Calculate total combinations
    for (int i = 0; i < n; i++) {
        int digit = digits[i] - '2';
        totalCombinations *= strlen(digitToChar[digit]);
    }

    // Allocate memory for result
    char **result = (char **)malloc(totalCombinations * sizeof(char *));
    char *combination = (char *)malloc((n + 1) * sizeof(char));
    combination[n] = '\0';

    backtrack(combination, 0, digits, result, returnSize);

    free(combination);
    return result;
}